syntax = "proto3";

package fedlearn.v1;

option java_package = "com.fedlearn.v1";
option java_multiple_files = true;


// The main service definition for the Federated Learning process.
service FederatedLearningService {
  // A client registers itself with the server to participate.
  rpc RegisterClient(RegisterClientRequest) returns (RegisterClientResponse);

  // A client requests the latest global model and training instructions.
  rpc GetGlobalModel(GetGlobalModelRequest) returns (GetGlobalModelResponse);

  rpc GetGlobalModelStream(GetGlobalModelRequest) returns (stream ModelChunk);
  // A client submits its trained model update to the server.
  rpc SubmitModelUpdate(SubmitModelUpdateReque) returns (SubmitModelUpdateResponse);


  rpc SubmitModelUpdateStream(stream ModelUpdateChunk) returns (SubmitModelUpdateResponse);

  // A client or administrator queries the server's current state.
  rpc GetServerStatus(GetServerStatusRequest) returns (GetServerStatusResponse);

  rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse);
}

// === Core Data Structures ===

message Tensor {
  bytes data = 1;         // Serialized raw tensor data.
  repeated int64 dims = 2;  // Shape of the tensor, e.g., [10, 784].
  string dtype = 3;       // Data type, e.g., "float32".
}

message ModelParameters {
  // A map of layer name to tensor for the entire model.
  map<string, Tensor> tensors = 1;
  // Metadata about the data used to generate these parameters.
  int64 num_examples_trained = 2;
}


// === RPC Message Definitions ===

// --- Registration ---
message RegisterClientRequest {
  string client_id = 1;
}

message RegisterClientResponse {
  enum Status {
    STATUS_UNSPECIFIED = 0;
    ACCEPTED = 1;
    REJECTED = 2;
  }
  Status status = 1;
  string message = 2;
}


// --- Model Distribution ---
message GetGlobalModelRequest {
  string client_id = 1;
}

message GetGlobalModelResponse {
  ModelParameters parameters = 1;
  int32 current_round = 2;
  map<string, string> config = 3; // e.g., {"local_epochs": "5", "batch_size": "64"}
}


// --- Model Update Submission ---
message SubmitModelUpdateReque {
  string client_id = 1;
  ModelParameters parameters = 2;
  int32 trained_on_round = 3; // Client reports which round it trained.
}

message SubmitModelUpdateResponse {
  // A simple acknowledgment.
  bool received = 1;
}


message GetServerStatusRequest {
}

message GetServerStatusResponse {
  enum ServerState {
    STATE_UNSPECIFIED = 0;
    INITIALIZING = 1;
    WAITING_FOR_CLIENTS = 2;
    AGGREGATING = 3;
    TRAINING_COMPLETE = 4;
  }
  ServerState server_state = 1;
  int32 current_round = 2;
  int32 required_clients_for_round = 3;
  int32 received_updates_this_round = 4;
}

message HeartbeatRequest {
    string client_id = 1;
    string status = 2;
    int32 current_step = 3;
    int32 total_steps = 4;
    int32 current_round = 5;
}

message HeartbeatResponse {
    bool acknowledged = 1;
    bool should_stop = 2;
    string message = 3;
}

message ModelUpdateChunk {
    string client_id = 1;
    int32 trained_on_round = 2;
    int32 chunk_index = 3;
    int32 total_chunks = 4;
    bytes chunk_data = 5;
    bool is_final_chunk = 6;
    int64 num_examples = 7;
}

message ModelChunk {
    int32 chunk_index = 1;
    int32 total_chunks = 2;
    bytes chunk_data = 3;
    bool is_final_chunk = 4;
    int32 current_round = 5;
    map<string, string> config = 6;
}